### ISSUE TITLE
The loss of BatchFormer with MoCo-v3 ?

### ISSUE THREAD

**Issue by iorange-77** [zhihou7/BatchFormer#18]
May I ask the 133-134 lines of code in ./moco-v3/moco/builder.py:
loss = self.contrastive_loss(q1[:N], k2[:N]) + self.contrastive_loss(q1[:N], k2[N:]) + self.contrastive_loss(q1[N:], k2[:N]) + self.contrastive_loss(q1[N:], k2[:N]) + \
                  self.contrastive_loss(q2[:N], k1[:N]) + self.contrastive_loss(q2[:N], k1[N:]) + self.contrastive_loss(q2[:N], k1[:N]) + self.contrastive_loss(q2[:N], k1[N:])
 
Is the loss calculated incorrectly? should be: ？？
loss = self.contrastive_loss(q1[:N], k2[:N]) + self.contrastive_loss(q1[:N], k2[N:]) + self.contrastive_loss(q1[N:], k2[:N]) + self.contrastive_loss(q1[N:], k2[N:]) + \
                  self.contrastive_loss(q2[:N], k1[:N]) + self.contrastive_loss(q2[:N], k1[N:]) + self.contrastive_loss(q2[N:], k1[:N]) + self.contrastive_loss(q2[N:], k1[N:])

If my understanding is wrong, what does this loss calculation method mean?

**Comment by zhihou7** [comment]
Dear @iorange-1014,
Many Thanks to you! It is a typo. It should be self.contrastive_loss(q2[N:], k1[:N]) + self.contrastive_loss(q2[N:], k1[N:]) in the last! I have updated the code.

Sincerely yours,

**Comment by zhihou7** [comment]
Hi @iorange-1014,
Thanks for your comment. Sorry for the terrible code. It seems like the model achieves clearly better performance with the revised code. The pretrained model is provided in [here](https://unisydneyedu-my.sharepoint.com/personal/zhou9878_uni_sydney_edu_au/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fzhou9878%5Funi%5Fsydney%5Fedu%5Fau%2FDocuments%2Freleased%5Fmodels%2Fmoco%2Dv3%5Fg34%2Fr50%5Ffixed%5Fmocov3%5Fe100checkpoint%2Epth%2Etar&parent=%2Fpersonal%2Fzhou9878%5Funi%5Fsydney%5Fedu%5Fau%2FDocuments%2Freleased%5Fmodels%2Fmoco%2Dv3%5Fg34&ga=1). Actually, I still think current losses might include redundant terms. I am not familiar with moco. And I do not have enough computational resources to tune this, but providing a naive and intuitive version.  Lastly, many thanks to you.

Regards,
